adapters/zmaticoo/zmaticoo.go

```diff
@@ -0,0 +1,189 @@
+			var nativeRequest map[string]interface{}
+			nativeCopyRequest := make(map[string]interface{})
+			err := json.Unmarshal([]byte(request.Imp[i].Native.Request), &nativeRequest)
+			//just ignore the bad native request
+			if err == nil {
+				_, exists := nativeRequest["native"]
+				if exists {
+					continue
+				}
+
+				nativeCopyRequest["native"] = nativeRequest
+				nativeReqByte, err := json.Marshal(nativeCopyRequest)
+				//just ignore the bad native request
+				if err != nil {
+					continue
+				}
+
+				nativeCopy := *request.Imp[i].Native
+				nativeCopy.Request = string(nativeReqByte)
+				request.Imp[i].Native = &nativeCopy
+			}
+		}
+	}
+}
+
+func getZmaticooExt(request *openrtb2.BidRequest) (*openrtb_ext.ExtImpZmaticoo, []error) {
+	var extImpZmaticoo openrtb_ext.ExtImpZmaticoo
+	var errs []error
+
+	for _, imp := range request.Imp {
+		var extBidder adapters.ExtImpBidder
+		err := json.Unmarshal(imp.Ext, &extBidder)
+		if err != nil {
+			errs = append(errs, err)
+			continue
+		}
+		err = json.Unmarshal(extBidder.Bidder, &extImpZmaticoo)
+		if err != nil {
+			errs = append(errs, err)
+			continue
+		}
+		break
+	}
+
+	return &extImpZmaticoo, errs
+
+}
+
+func (a *adapter) getEndpoint(ext *openrtb_ext.ExtImpZmaticoo) (string, error) {
+	return macros.ResolveMacros(a.EndpointTemplate, macros.EndpointTemplateParams{Host: "gw-" + url.QueryEscape(ext.ZoneId) + "-bid.zmaticoo.com"})
+}
+
+// MakeBids make the bids for the bid response.
+func (a *adapter) MakeBids(internalRequest *openrtb2.BidRequest, externalRequest *adapters.RequestData, response *adapters.ResponseData) (*adapters.BidderResponse, []error) {
+	if response.StatusCode == http.StatusNoContent {
+		return nil, nil
+	}
+
+	if response.StatusCode == http.StatusBadRequest {
+		return nil, []error{&errortypes.BadInput{
+			Message: fmt.Sprintf("Unexpected status code: %d.", response.StatusCode),
+		}}
+	}
+
+	if response.StatusCode != http.StatusOK {
+		return nil, []error{&errortypes.BadServerResponse{
+			Message: fmt.Sprintf("Unexpected status code: %d.", response.StatusCode),
+		}}
+	}
+
+	var bidResp openrtb2.BidResponse
+
+	if err := json.Unmarshal(response.Body, &bidResp); err != nil {
+		return nil, []error{err}
+	}
+
+	bidResponse := adapters.NewBidderResponseWithBidsCapacity(1)
+
+	for _, sb := range bidResp.SeatBid {
+		for i := range sb.Bid {
+			var mediaType = getBidType(sb.Bid[i].ImpID, internalRequest.Imp)
+			bidResponse.Bids = append(bidResponse.Bids, &adapters.TypedBid{
+				Bid:     &sb.Bid[i],
+				BidType: mediaType,
+			})
+		}
+	}
+	return bidResponse, nil
+
+}
+
+func getBidType(impId string, imps []openrtb2.Imp) openrtb_ext.BidType {
+	bidType := openrtb_ext.BidTypeBanner
+	for _, imp := range imps {
+		if imp.ID == impId {
+			if imp.Banner != nil {
+				break
+			}
+			if imp.Video != nil {
+				bidType = openrtb_ext.BidTypeVideo
+				break
+			}
+			if imp.Native != nil {
+				bidType = openrtb_ext.BidTypeNative
+				break
+			}
+
+		}
+	}
```

Comment:
Warning - This is an anti-pattern. This code assumes that for multi-format bid request, mediatype will be set to type banner by default. I would recommend to set the mediaType based on bidder response. Your bidder should be able to set openrtb2.Bid.MType field and use that field here to get mediaTypeForImp. Even if your bidder does not support multi-format request, I would highly encourage you to make this change.

Reply:
ok,thank you ,i will change